### ISSUE TITLE
Move ProxImaL bindings and examples to `contrib`

### ISSUE THREAD

**Issue by kohr-h** [odlgroup/odl#1169]
This is nice to have but not really within the scope of core ODL. It also brings issues with testing (need to exclude files etc.). I think we should move it to `contrib`.

**Comment by adler-j** [comment]
Good suggestion. Same applies to stir i guess

**Comment by ozanoktem** [comment]
Agree to some extent with @kohr-h and @adler-j . On the other hand, why do we then treat ASTRA differently? Or sci-kit image for that matter? We apparently consider bindings to selected external libraries as an essential part of ODL, e.g. the numpy library is indeed absolutely necessary for ODL to work. On the other hand, ASTRA is not necessary at this low level. DO we consider ASTRA to be a  core library that should be part of ODL?

**Comment by adler-j** [comment]
Basically the ASTRA bindings are mature, while the stir bindings look like they were written by a 14 year old as a high school project. Once STIR is properly cleaned up I guess it can come back.

**Comment by ozanoktem** [comment]
I understand this, but perhaps we need a criteria that is less subjective than saying that code quality is inferior and seems to be "written by a 14 year old as a high school project". As an example, are the ProximaL bindings also of such low quality that they can be considered as being "written by a 14 year old as a high school project"? 

Clearly we have a notion of core functionality, which is the functional analysis part of ODL. This requires numpy and some other stuff related to handling the operator calculus and discretizations, but not necessarily handling tomographic geometries, computing retransforms or Fourier transforms, etc. 

The next level are "basic operators" that we consider important for many applications. It is here that we include bindings to ASTRA, pyFFT/FFTW, pyWavelets, STIR, ,.... Clearly, a binding must be coded in a sufficiently generic manner with tests in order for it to qualify as a core operator.  This is the case for ASTRA which made it necessary to establish the entire tomographic geometry package. It is also the case (I hope) for the Fourier transform (using pyFFT/FFTW) and to some extent, the various wavelet transforms (using pyWavelets). It is obviously not the case for PET using STIR, e.g. we lack data structures in ODL for representing PET geometries. 

So, should these distinctions also be reflected in how we organise the ODL Git repo?

**Comment by adler-j** [comment]
> I understand this, but perhaps we need a criteria that is less subjective than saying that code quality is inferior and seems to be "written by a 14 year old as a high school project". As an example, if we are to move ProximaL to contrib, is it because the ProximaL bindings also of such low quality that they can be considered as being written by a 14 year old as a high school project? I guess not, right?

Basically yes. I wrote them once as a test and saw that they ran. They have basically no actual use so far and we barely know if they work with the latest proximal versions. The code, as currently written, would frankly have a hard time getting accepted if it was proposed right now. The main issue is perhaps not the code quality per-se, but that it does not have a context and is basically "hanging code".

> ... So, should these distinctions also be reflected in how we organise the ODL Git repo?

Yes, we actually have a quite specific description of what should go into [contrib]( https://github.com/odlgroup/odl/tree/master/odl/contrib), specifically:

> This is a special place for user-contributed modules. Its intention is to provide a "fast lane" for you as a contributor to include code that is:
>
> * outside the scope of the main ODL library,
> * experimental and untested,
> * still subject to major breaking changes or
> * "not quite there yet" in terms of documentation, testing or completeness.

and we follow this quite well, except perhaps for STIR and proximal. Is there anything with these guidelines that you propose changing?

If this is not a contrib vs not-contrib issue, we've had quite some talks on creating a "core" library, but the problem with this is that it basically means moving some of the content into yet another sub-folder, e.g. move `odl.space` into `odl.core.space`, and thats it. As a user, anything available under `odl.my_feature` is kindoff implicitly "core".

**Comment by kohr-h** [comment]
Wow, this kicked off quite some discussion. I wrote the guidelines for `contrib` packages, and I am aware that all of the points are subjective to some extent. 

One important point that we should probably add to the list is that **it requires a third-party package**. Now, that applies to ASTRA, pyFFTW and PyWavelets, too, **but** those are mature packages that are right within the core scope of ODL.

For ProxImaL and STIR, I have different reasons for thinking that they should move to `contrib`.

### ProxImaL
**Scope:** The *purpose* is in ODL scope (access to convex optimization methods), but since the code is usually quite slow, the addition to ODL is limited IMO.
**Maturity:** Not so much as far as I know.
**3rd-party packages:** Obviously introduces dependency on `proximal`, but it's available on PyPI so that's no problem.

In summary I see it more like a nice small addition that's fun to play with, but not so important.

### STIR
**Scope:** Fits the scope very well since it enables nuclear imaging, **but** it's slow and somewhat arcane, so I'd say it's better than nothing but certainly not great.
**Maturity:** Good. Interoperability seems to be an issue though.
**3rd-party packages:** This is a **huge** issue because you need to compile the source yourself, which is a big pain in the butt.

Here my conclusion is that with proper Python packaging and better bindings we can reconsider it, but for now I'm clearly for moving it to `contrib` (basically due to the "not there yet" reason).

**Comment by adler-j** [comment]
Gonna get this fixed, any suggestion on where we should move the proximal stuff? Feels weird to have its own library for one function

**Comment by kohr-h** [comment]
I guess that's the only sensible solution. Putting it into another place like `contrib/solvers` would make that part require `proximal` to be installed, which would be weird.
